pkg/dynamiclifecycle/reconciler.go

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package dynamiclifecycle
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"iter"
+	"log/slog"
+
+	"github.com/cilium/statedb"
+	"github.com/cilium/statedb/reconciler"
+	"k8s.io/utils/clock"
+
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+type ops struct {
+	clock clock.Clock
+	log   *slog.Logger
+	t     statedb.Table[TableLifecycle]
+	db    *statedb.DB
+}
+
+// Update checks the TableLifecycle enablement flag.
+// If DynamicFeature is disabled, the lifecycle is stopped.
+// Currently, there is no dependency checking when stopping feature.
+// If DynamicFeature is enabled, the list of dependencies is checked to ensure
+// all dependants are running and if the requirement is met the lifecycle is started.
+func (o *ops) Update(ctx context.Context, txn statedb.ReadTxn, tl TableLifecycle) error {
+	// Nothing to do
+	if tl.Enabled == tl.FeatureLifecycle.IsRunning() {
+		return nil
+	}
+
+	if !tl.Enabled {
+		return tl.FeatureLifecycle.Stop(o.log, ctx)
+	}
+
+	if err := o.checkDependencies(tl.FeatureLifecycle.GetDependencies(), true); err != nil {
+		return err
+	}
+
+	tl.StartedAt = o.clock.Now()
+	err := tl.FeatureLifecycle.Start(o.log, ctx)
+	tl.StartDuration = o.clock.Since(tl.StartedAt)
+
+	if err != nil {
+		o.log.Error("Failed to start all hooks, stopping started hooks", "feature", tl.FeatureLifecycle.GetName(), logfields.Error, err)
+		errStop := tl.FeatureLifecycle.Stop(o.log, ctx)
+		return errors.Join(err, errStop)
+	}
+
+	return nil
+}
+
+// Delete stops the DynamicFeature if it's enabled or running and the reconciler status is Done or Error.
+// Currently, no dependencies are checked on stopping the DynamicFeature.
+func (o *ops) Delete(ctx context.Context, txn statedb.ReadTxn, tl TableLifecycle) error {
+	// Nothing to do
+	if tl.Enabled == tl.FeatureLifecycle.IsRunning() {
+		return nil
+	}
+
+	enabledOrRunning := tl.Enabled || tl.FeatureLifecycle.IsRunning()
+	statusDoneOrError := tl.Status.Kind == reconciler.StatusKindDone || tl.Status.Kind == reconciler.StatusKindError
+
+	if enabledOrRunning && statusDoneOrError {
+		return tl.FeatureLifecycle.Stop(o.log, ctx)
+	}
+	return nil
+}
+
+func (o *ops) Prune(_ context.Context, _ statedb.ReadTxn, _ iter.Seq2[TableLifecycle, statedb.Revision]) error {
+	return nil
+}
+
+func (o *ops) checkDependencies(dependencies []DynamicFeature, expectEnabled bool) error {
+	readTxn := o.db.ReadTxn()
```

Comment:
Don't do a ReadTxn inside the ops as this is racy! Update is called with a ReadTxn so pass that in here. 

The reason it is racy is because Update() is called based on the state of status in a snapshot and it might've changed so we might read some future state here, which may or may not get us in trouble. Better to just use the one ReadTxn for everything to avoid issues.